<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-данных-из-файла" data-toc-modified-id="Изучение-данных-из-файла-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение данных из файла</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Результат" data-toc-modified-id="Результат-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Разбиение-данных-на-выборки" data-toc-modified-id="Разбиение-данных-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбиение данных на выборки</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Исследование-моделей" data-toc-modified-id="Исследование-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование моделей</a></span><ul class="toc-item"><li><span><a href="#Дерево-принятия-решений-(Decision-Tree-Classifier)" data-toc-modified-id="Дерево-принятия-решений-(Decision-Tree-Classifier)-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Дерево принятия решений (Decision Tree Classifier)</a></span></li><li><span><a href="#Результат" data-toc-modified-id="Результат-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Результат</a></span></li><li><span><a href="#Случайный-лес-(Random-Forest-Classifier)" data-toc-modified-id="Случайный-лес-(Random-Forest-Classifier)-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Случайный лес (Random Forest Classifier)</a></span></li><li><span><a href="#Результат" data-toc-modified-id="Результат-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Результат</a></span></li><li><span><a href="#Логистическая-регрессия-(Logistic-Regression)" data-toc-modified-id="Логистическая-регрессия-(Logistic-Regression)-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Логистическая регрессия (Logistic Regression)</a></span></li><li><span><a href="#Результат" data-toc-modified-id="Результат-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Результат</a></span></li><li><span><a href="#Сравнение-моделей" data-toc-modified-id="Сравнение-моделей-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Сравнение моделей</a></span></li><li><span><a href="#Результат" data-toc-modified-id="Результат-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Проверка-модели-на-тесторовой-выборке" data-toc-modified-id="Проверка-модели-на-тесторовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка модели на тесторовой выборке</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Проверка-модели-на-адекватность" data-toc-modified-id="Проверка-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка модели на адекватность</a></span><ul class="toc-item"><li><span><a href="#Результат" data-toc-modified-id="Результат-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Результат</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Описание проекта

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».
В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.  
Постройте модель с максимально большим значением accuracy. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте accuracy на тестовой выборке самостоятельно.
 

# Описание данных

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:  
•	сalls — количество звонков,  
•	minutes — суммарная длительность звонков в минутах,  
•	messages — количество sms-сообщений,  
•	mb_used — израсходованный интернет-трафик в Мб,   
•	is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

 

## Изучение данных из файла

In [1]:
#Импорт библиотек
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import matplotlib.style
import matplotlib as mpl

import sklearn 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import plot_tree

from sklearn.dummy import DummyClassifier

import warnings
warnings.filterwarnings("ignore")

pd.set_option('display.max_colwidth', -1)

In [2]:
import os

pth1 = 'D:/YANDEX.PRAKTIKUM/DATA_SCIENCE/Введение_в_машинное_обучение/users_behavior.csv'
pth2 = '/datasets/users_behavior.csv'

if os.path.exists(pth1):
    UB = pd.read_csv(pth1)
elif os.path.exists(pth2):
    UB = pd.read_csv(pth2)
else:
    print('Something is wrong')

In [3]:
#Считаем датасет
print(UB.info())
display(UB.describe())
UB.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None


calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000  3214.000000   3214.000000
mean   63.038892    438.208787   38.281269    17207.673836  0.306472   
std    33.236368    234.569872   36.148326    7570.968246   0.461100   
min    0.000000     0.000000     0.000000     0.000000      0.000000   
25%    40.000000    274.575000   9.000000     12491.902500  0.000000   
50%    62.000000    430.600000   30.000000    16943.235000  0.000000   
75%    82.000000    571.927500   57.000000    21424.700000  1.000000   
max    244.000000   1632.060000  224.000000   49745.730000  1.000000

calls  minutes  messages   mb_used  is_ultra
0  40.0   311.90   83.0      19915.42  0       
1  85.0   516.75   56.0      22696.96  0       
2  77.0   467.66   86.0      21060.45  0       
3  106.0  745.53   81.0      8437.39   1       
4  66.0   418.74   1.0       14502.75  0

### Вывод

Датасет состоит из 3214 наблюдений и 5 столбцов. Предобработка данных уже выполнена, поэтому пропусков нет.

Поскольку стобец is_ultra содержит объекты булева типа («Ультра» — 1, «Смарт» — 0), поэтому заменим тип данных для этого признака на логический или булев тип.

In [4]:
UB['is_ultra'] = UB['is_ultra'].astype('bool')
UB.dtypes

calls       float64
minutes     float64
messages    float64
mb_used     float64
is_ultra    bool   
dtype: object

Посмотрим, сколько и какая долюя объектов каждого типа — «Ультра» и «Смарт».

In [5]:
values = UB['is_ultra'].value_counts()
values = values.to_frame().reset_index()
values.columns = ['Boolean', 'Count']
values['Percentage'] = UB['is_ultra'].value_counts(normalize=True)
pd.DataFrame(values)

values['Tariff'] = values['Boolean'].apply(lambda x: 'Ультра' if x == True else 'Смарт')
values

Boolean  Count  Percentage  Tariff
0  False    2229   0.693528    Смарт 
1  True     985    0.306472    Ультра

Расчеты показали, что 69% пользователей используют тариф «Смарт», а 31% — тариф «Ультра».

### Результат

После открытия и изучения данных, провели замену типа данных для признака is_ultra и еще раз убедились, что данные не содержат пропуски.

## Разбиение данных на выборки

Разделим исходные данные на три выборки: обучающую, валидационную и тестовую.  
Сначала методом разделим исходные данные на обучающую (60%) и валидационную выборку (40%). После этого разделим валидационную выборку пополам — на валидационную и тестовую выборки. Таким образом каждая из этих 2-х выборок составит 20% из всех данных датафрейма.

In [6]:
UB_train, UB_valid = train_test_split(UB, train_size=0.60, test_size=0.40, random_state=123)
UB_valid, UB_test = train_test_split(UB_valid, train_size=0.50, test_size=0.50, random_state=123)

Проверим.

In [7]:
samples = {'Обучающая выборка' : UB_train, 
           'Валидационная выборка': UB_valid, 
           'Тестовая выборка': UB_test}

for key, value in samples.items():
    print(key + ':', round(len(value) / len(UB)*100), '%')

Обучающая выборка: 60 %
Валидационная выборка: 20 %
Тестовая выборка: 20 %


В каждой выборке выделим признаки и целевой признак. В данном случае целевым признаком является значения столбца is_ultra, а признаками — значение всех остальных столбцов датафрейма UB.

In [8]:
# обучающая выборка
features_train = UB_train.drop(['is_ultra'], axis=1)
target_train = UB_train['is_ultra']

# валидационная выборка
features_valid = UB_valid.drop(['is_ultra'], axis=1)
target_valid = UB_valid['is_ultra']

# тестовая выборка
features_test = UB_test.drop(['is_ultra'], axis=1)
target_test = UB_test['is_ultra']

### Результат

Разбили исходные данные на обучающую, валидационную и тестовую выборки, и в каждой из них выделили признаки и целевой признак.

## Исследование моделей

Рассмотрим следующие модели классификации:    
    - дерево принятия решений (Decision Tree Classifier)  
    - случайный лес (Random Forest Classifier)  
    - логистическая регрессия (Logistic Regression).

### Дерево принятия решений (Decision Tree Classifier)

Проверим результат модели без смены гиперпараметров.

In [9]:
model_dtc = DecisionTreeClassifier(random_state=123)
model_dtc.fit(features_train, target_train)
predictions_valid = model_dtc.predict(features_valid)
accuracy_dtc = accuracy_score(target_valid, predictions_valid)

print('Result:', accuracy_dtc)

Result: 0.7465007776049767


Сохраним результат в переменной results_from_initial_models, чтобы в итоге можно было сравнить, какая модель показала наилучший результат без настройки гиперпараметров

In [10]:
results_from_initial_models = []
results_from_initial_models.append({'DecisionTreeClassifier': accuracy_dtc})
results_from_initial_models

[{'DecisionTreeClassifier': 0.7465007776049767}]

Проведем перебор гиперпараметра max-depth (глубина деревьев) от  до 11 чтобы посмотреть улучшит ли это работу модели. Результаты сохраним в переменной results_dtc и выведем результат.

In [11]:
results_dtc = []

for depth in range(1,11):
    model = DecisionTreeClassifier(random_state=123, max_depth=depth)
    model.fit(features_train, target_train)
    
    predictions_valid = model.predict(features_valid)
    accuracy_valid = accuracy_score(target_valid, predictions_valid)

    results_dtc.append({'Model name': 'DecisionTreeClassifier', 
                    'Hyperparameters': {'random_state': 123, 'max_depth':depth}, 
                    'Accuracy score': accuracy_valid})
UB_dtc = pd.DataFrame.from_dict(results_dtc)
print(UB_dtc)

               Model name                         Hyperparameters  \
0  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 1}    
1  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 2}    
2  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 3}    
3  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 4}    
4  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 5}    
5  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 6}    
6  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 7}    
7  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 8}    
8  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 9}    
9  DecisionTreeClassifier  {'random_state': 123, 'max_depth': 10}   

   Accuracy score  
0  0.762053        
1  0.790047        
2  0.802488        
3  0.804044        
4  0.822706        
5  0.814930        
6  0.811820        
7  0.802488        
8  0.818040        
9  0.802488        


### Результат

Наилучший результат модель дает при глубине 5 (accuracy_score = ~0,822706).

Самый лучший результат сохраним в переменной best_models_results.

In [12]:
best_models_results = []
best_models_results.append(UB_dtc.loc[4])

### Случайный лес (Random Forest Classifier)

Проверим результат модели без смены гиперпараметров. Сохраним результат в переменной results_from_initial_models.

In [13]:
model_rfc = RandomForestClassifier(random_state=123) 
model_rfc.fit(features_train, target_train)  

predictions_rfc = model_rfc.predict(features_valid) 
accuracy_rfc = accuracy_score(target_valid, predictions_rfc)

print('Result:', accuracy_rfc)

Result: 0.8180404354587869


In [14]:
results_from_initial_models.append({'RandomForestClassifier': accuracy_rfc})
results_from_initial_models

[{'DecisionTreeClassifier': 0.7465007776049767},
 {'RandomForestClassifier': 0.8180404354587869}]

Результат у модели случайного леса без смены гиперпараметров лучше чем у дерева принятия решений. 

Настроим гиперпараметры n_estimators (оличество деревьев в "лесу") и max_depth (глубину деревьев). Результат запишем в results_rfc. Выведем на экран комбинацию гиперпараметров, которая дает самый высокий показатель аccuracy_score.

In [15]:
results_rfc = []

for depth in range(1,11):
    
    for estimator in range(10, 101, 10):
        
        model = RandomForestClassifier(random_state=123, n_estimators=estimator, max_depth=depth) 
        model.fit(features_train, target_train)
    
        predictions_valid = model.predict(features_valid)
        accuracy_valid = accuracy_score(target_valid, predictions_valid)
       
        results_rfc.append({'Model name': 'RandomForestClassifier', 
                            'Hyperparameters': {'random_state': 123, 
                                                'n_estimators': estimator, 
                                                'max_depth':depth}, 
                            'Accuracy score': accuracy_valid})
UB_rfc = pd.DataFrame.from_dict(results_rfc)
UB_rfc[UB_rfc['Accuracy score']==UB_rfc['Accuracy score'].max()]        

Model name  \
86  RandomForestClassifier   

                                              Hyperparameters  Accuracy score  
86  {'random_state': 123, 'n_estimators': 70, 'max_depth': 9}  0.835148

### Результат

Наилучший результат модель дает при гиперпараметре n_estimators равном 70, а max_depth = 9 (accuracy_score = ~0.835148).

Самый лучший результат сохраним в переменной best_models_results.

In [16]:
best_models_results.append(UB_rfc.loc[86])

### Логистическая регрессия (Logistic Regression)

Проверим результат модели без смены гиперпараметров. Сохраним результат в переменной results_from_initial_models.

In [17]:
model_lr = LogisticRegression() 
model_lr.fit(features_train, target_train)
predictions_lr = model_lr.predict(features_valid)

accuracy_lr = accuracy_score(target_valid, predictions_lr)
accuracy_lr

0.7589424572317263

In [18]:
results_from_initial_models.append({'LogisticRegression': accuracy_lr})
results_from_initial_models

[{'DecisionTreeClassifier': 0.7465007776049767},
 {'RandomForestClassifier': 0.8180404354587869},
 {'LogisticRegression': 0.7589424572317263}]

Результат у модели логистической регрессии без смены гиперпараметров не превзошел наилучший результат модели случайного леса. 

In [19]:
results_lr = []

model_lr = LogisticRegression(random_state=123, solver='lbfgs', penalty='l2') 
model_lr.fit(features_train, target_train)
predictions_lr = model_lr.predict(features_valid)

accuracy_lr = accuracy_score(target_valid, predictions_lr)

results_lr.append({'Model name': 'LogisticRegression', 
                   'Hyperparameters': {'random_state': 123, 'solver':'lbfgs', 'penalty': 'l2'}, 
                   'Accuracy score': accuracy_lr})

accuracy_lr

0.7589424572317263

При первой смене гиперпараметров изменений не произошло. 

Поменяем значение гиперпараметра penalty на l1, а гиперпараметра solver на liblinear.

In [20]:
model_lr = LogisticRegression(random_state=123, solver='liblinear', penalty='l1')  
model_lr.fit(features_train, target_train)
predictions_lr = model_lr.predict(features_valid)

accuracy_lr = accuracy_score(target_valid, predictions_lr)

results_lr.append({'Model name': 'LogisticRegression', 
                   'Hyperparameters': {'random_state': 123, 'solver':'liblinear', 'penalty': 'l1'}, 
                   'Accuracy score': accuracy_lr})

accuracy_lr

0.76049766718507

### Результат

Показатель accuracy_score при данной комбинации гиперпараметров стал лучше. 

Самый лучший результат сохраним в переменной best_models_results.

In [21]:
UB_lr = pd.DataFrame.from_dict(results_lr)
best_models_results.append(UB_lr.loc[1])

### Сравнение моделей

Выведем, какие результаты дает каждая из моделей без настройки гиперпараметров.

In [22]:
pd.DataFrame(results_from_initial_models) 

DecisionTreeClassifier  RandomForestClassifier  LogisticRegression
0  0.746501               NaN                     NaN                
1 NaN                      0.81804                NaN                
2 NaN                     NaN                      0.758942

Самый лучший результат accuracy_score показывает модель случайного леса (0.81804), а самый худший — дерево принятия решений (0.746501).

Выведем ранее отобранные модели, которые показали самый лучший результат в каждой из модели. Из них выберем лучшую и протестируем ее на тестовой выборке.

In [25]:
pd.DataFrame(best_models_results)

Model name  \
4   DecisionTreeClassifier   
86  RandomForestClassifier   
1   LogisticRegression       

                                                  Hyperparameters  \
4   {'random_state': 123, 'max_depth': 5}                           
86  {'random_state': 123, 'n_estimators': 70, 'max_depth': 9}       
1   {'random_state': 123, 'solver': 'liblinear', 'penalty': 'l1'}   

    Accuracy score  
4   0.822706        
86  0.835148        
1   0.760498

Наилучший результат показала модель случайного леса с гиперпараметрами n_estimators = 70 и max_depth = 9. Укаазанная модель подлжеит тестированию на тестовой выборке.

### Результат

Исследовано три модели классификации: дерево принятия решений, случайный лес и логистическая регрессия.  
Каждая из моделей обучена на обучающей выборке и проверена на валидационной выборке.  
Наилучший результат показала модель случайного леса (показатель accuracy_score без настройки гиперпараметров = 0.81804, с настройкой гиперпараметров = 0.835148).

## Проверка модели на тесторовой выборке

Проверим модель случайного леса с гиперпараметрами n_estimators = 70 и max_depth = 9 на тестовой выборке.

In [28]:
model_test_final = RandomForestClassifier(random_state=123, n_estimators=70, max_depth=9)
model_test_final.fit(features_train, target_train)  
predictions_test_final = model_test_final.predict(features_test)

accuracy_test_final = accuracy_score(target_test, predictions_test_final)
accuracy_test_final 

0.7869362363919129

### Результат

Модель работает достаточно хорошо — она способна правильно предсказать около 78% ответов.

## Проверка модели на адекватность

Используем классификатор DummyClassifier из библиотеки Skelarn.
Выберем 3 стратегии — stratified, most_frequent и uniform. Если их способность предсказать результат будет лучше чем у настроенной модели случайного леса, значит модель работает плохо и ее необходимо настраивать заного.
    
    
    
    


In [29]:
strategies = ['stratified', 'most_frequent', 'uniform'] 
  
dummy_models_results = [] 
for strategy in strategies: 
    dc = DummyClassifier(strategy = strategy, random_state = 42)
    
    dc.fit(features_train, target_train) 
    result = dc.score(features_test, target_test) 
    dummy_models_results.append({strategy: result}) 

pd.DataFrame(dummy_models_results) 

stratified  most_frequent   uniform
0  0.578538   NaN            NaN      
1 NaN          0.693624      NaN      
2 NaN         NaN             0.475894

У финальной модели случайного леса показатель accuracy_score был 0.78693, что гораздо выше, чем те показатели, которые выдал классификатор DummyClassifier(). Таким образом, можно сделать вывод, что модель случайного леса работает достаточно хорошо и ее можно использовать на практике.

### Результат

Модель случайного леса с гиперпараметрами n_estimators = 70 и max_depth = 9 работает лучше, чем стратегии случайного прогнозирования классификатора DummyClassifier().

## Вывод

Проверка модели на тестовой выборке показала, что модель работает достаточно хорошо — она способна правильно предсказать около 78% ответов.  